Objects/tupleobject.c

```diff
@@ -333,39 +333,46 @@ tuplerepr(PyTupleObject *v)
     return NULL;
 }
 
-/* The addend 82520, was selected from the range(0, 1000000) for
-   generating the greatest number of prime multipliers for tuples
-   up to length eight:
 
-     1082527, 1165049, 1082531, 1165057, 1247581, 1330103, 1082533,
-     1330111, 1412633, 1165069, 1247599, 1495177, 1577699
+/* Hash for tuples. This is a slightly simplified version of the xxHash
+   non-cryptographic hash:
+   - we do not use any parallellism, there is only 1 accumulator.
+   - we drop the final mixing since this is just a permutation of the
+     output space: it does not help against collisions.
+   - at the end, we mangle the length with a single constant as a
+     "class signature" to make tuple hashes different from other hashes
+     based on xxHash.
+   For the xxHash specification, see
+   https://github.com/Cyan4973/xxHash/wiki/xxHash-specification-(draft)
 
    Tests have shown that it's not worth to cache the hash value, see
-   issue #9685.
+   https://bugs.python.org/issue9685
 */
-
 static Py_hash_t
 tuplehash(PyTupleObject *v)
 {
-    Py_uhash_t x;  /* Unsigned for defined overflow behavior. */
-    Py_hash_t y;
-    Py_ssize_t len = Py_SIZE(v);
-    PyObject **p;
-    Py_uhash_t mult = _PyHASH_MULTIPLIER;
-    x = 0x345678UL;
-    p = v->ob_item;
-    while (--len >= 0) {
-        y = PyObject_Hash(*p++);
-        if (y == -1)
+    Py_ssize_t i, len = Py_SIZE(v);
+    PyObject **item = v->ob_item;
+
+    Py_uhash_t acc = _PyHASH_XXPRIME_5;
+    for (i = 0; i < len; i++) {
+        Py_uhash_t lane = PyObject_Hash(item[i]);
+        if (lane == (Py_uhash_t)-1) {
             return -1;
-        x = (x ^ y) * mult;
-        /* the cast might truncate len; that doesn't change hash stability */
-        mult += (Py_hash_t)(82520UL + len + len);
+        }
+        acc += lane * _PyHASH_XXPRIME_2;
+        acc = _PyHASH_XXROTATE(acc);
+        acc *= _PyHASH_XXPRIME_1;
+    }
+
+    /* Add input length, mangled to provide a class signature. The constant
+       is chosen to keep the historical value of hash(()). */
+    acc += len ^ (_PyHASH_XXPRIME_5 ^ 3527539);
+
+    if (acc == (Py_uhash_t)-1) {
+        return -2;
```

Comment:
While we're at it, we could return a more interesting value than -2, something random and large.

Reply:
I don't see how returning a different value than -2 would help anything. If the hash function is sufficiently random (which we assume it is), then both the values -1 and -2 should be very rare. So replacing -1 by -2 increases collisions by a very very tiny amount. But replacing -1 by, say 1546275796, would increase the number of collisions by the same very very tiny amount.

And then I have to invent yet another random constant which I want to avoid if possible. It would just be additional complication without any gain.

Reply:
If you insist and come up with a concrete proposal for the random value, I'll make the change anyway.

Reply:
> If the hash function is sufficiently random (which we assume it is)

But we don't.  In particular, we know for sure that `hash(-1) == hash(-2) == -2`, and hashes of little integers are common as dirt.  While I doubt we'd ever be able to measure the difference, on the face of it it's attractive not to systematically create even more cases that always return -2.

> If you insist and come up with a concrete proposal for the random value, I'll make the change anyway.

You already did!  :smile: 1546275796 is fine (far less likely than -2 to show up as a hash, and fits in 31 bits).

Reply:
>  In particular, we know for sure that `hash(-1) == hash(-2) == -2`, and hashes of little integers are common as dirt. While I doubt we'd ever be able to measure the difference, on the face of it it's attractive not to systematically create even more cases that always return -2.

OK, I see your point: it's not about avoiding collisions between hashes of tuples, it's about avoiding collision between the hash of a tuple and the hash of a non-tuple (like the integer `-2`).

Reply:
Yup!  And it's an exceedingly weak point.  Nevertheless, the change can be made in far less time than it takes to argue about :wink:

Reply:
OK, done